library/pkg/executor/executor.go

```diff
@@ -0,0 +1,359 @@
+	}
+
+	if cgroupsV2Mounted() {
+		return nil
+	}
+
+	err = syscall.Mount("cgroup2", cgroupFolder, "cgroup2", 0, "")
+
+	if err != nil {
+		return fmt.Errorf("failed to mount cgroupv2 %w", err)
+	}
+
+	if cgroupsV2Mounted() {
+		return nil
+	}
+
+	return errors.New("failed to mount cgroups v2")
+}
+
+func cgroupsV2Mounted() bool {
+	mountCommand := exec.Command("mount", "-t", "cgroup2")
+	output, err := mountCommand.Output()
+
+	if err != nil {
+		return false
+	}
+	return len(output) != 0
+}
+
+func ensureCgroupsV1IsNotMounted() error {
+	mountCommand := exec.Command("mount", "-t", "cgroup")
+	output, err := mountCommand.Output()
+
+	if err != nil {
+		return fmt.Errorf("failed to check v1 cgroups, %w", err)
+	}
+
+	if len(output) == 0 {
+		return nil
+	}
+
+	for _, line := range strings.Split(string(output), "\n") {
+		path := strings.Split(line, " ")[2]
+		err = syscall.Unmount(path, 0)
+		if err != nil {
+			return fmt.Errorf("failed to unmount v1 cgroup at %s, %w", path, err)
+		}
+	}
+
+	return nil
+}
+
+// StartJob will execute the supplied command with the supplied args
+// processId is used to create the underlying cgroup for the process to be executed in
+//
+// Before the job is started the following restrictions will be set in cgroups:
+// * cpu.max set based on the CpuPercent of the Config of the Executor
+// * memory.max set based on the MemoryBytes of the Config of the Executor
+// * memory.high set based on the MemoryBytes * 0.75 of the Config of the Executor
+// * io.max set based on the WriteBytePerSec and ReadBytePerSec of the Config of the Executor for all disks
+//
+// in case there is any problem with the setup of the process, these cgroup controls cannot be created an error will be returned
+// In case the process setup was successful a Command will be returned
+// The Command will either have "running" or "failed" state based on whether the process was succesfully started
+// Further interaction with the process should happen through the returned Command instance
+func (e Executor) StartJob(processId string, command string, args []string) (Command, error) {
+	err := e.createCgroup(processId)
+
+	if err != nil {
+		cleanCgroups(processId)
+		return nil, err
+	}
+
+	cmd := exec.Command(command, args...)
+
+	outputBuffer := &internal.MultiplexingBuffer{}
+
+	cmd.Stderr = os.Stderr
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = outputBuffer
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Cloneflags: syscall.CLONE_NEWUTS | syscall.CLONE_NEWPID | syscall.CLONE_NEWUSER | syscall.CLONE_NEWNET | syscall.CLONE_NEWNS,
+		UidMappings: []syscall.SysProcIDMap{
+			{
+				ContainerID: 0,
+				HostID:      os.Getuid(),
+				Size:        1,
+			},
+		},
+		GidMappings: []syscall.SysProcIDMap{
+			{
+				ContainerID: 0,
+				HostID:      os.Getgid(),
+				Size:        1,
+			},
+		},
+	}
+
+	err = cmd.Start()
+
+	if err != nil {
+		return &internal.FailedCommand{}, nil
+	}
+
+	pid := cmd.Process.Pid
+
+	err = e.addPidToCgroups(processId, pid)
```

Comment:
What happens if the job started above spawns a child before it got added to the cgroup?

Reply:
This is something I mentioned in the tradeoff section of the design doc:
![image](https://user-images.githubusercontent.com/5039095/127623548-fe8f8c70-f0c9-4f14-b216-16d9eba27807.png)

But I can do some research how to do it better